src/wallet/coin_selection.rs

```diff
@@ -446,10 +446,12 @@ impl<D: Database> CoinSelectionAlgorithm<D> for BranchAndBoundCoinSelection {
             .map(|u| OutputGroup::new(u, fee_rate))
             .collect();
 
-        // Mapping every (UTXO, usize) to an output group.
+        // Mapping every (UTXO, usize) to an output group, filtering UTXOs with a negative
+        // effective value
         let optional_utxos: Vec<OutputGroup> = optional_utxos
             .into_iter()
             .map(|u| OutputGroup::new(u, fee_rate))
+            .filter(|u| u.effective_value.is_positive())
```

Comment:
I think this is fine to compute the `curr_value` but there is a reason to still include the `output_groups` with negative effective value in the vector consumed by `bnb` and `single_random_draw`. E.g.:
There are three utxos in the wallet:
(A) with effective value of 100
(B) with effective value of 100
(C) with effective value of -75
If `actual_target` is equal to 125, wouldn't be `Vec<A,B,C>` the best selection possible for `bnb` or `single_random_draw`?

Reply:
I don't think so, with your solution you would spend all three UTXOs and not create a change. The final balance of the wallet would be zero.

If you exclude `C`, then the coin selection would pick `A + B` and create a change so at the end the balance of your wallet is `C` + the change. Which is arguably a better outcome for the user.

Remember that the effective value isn't fixed for a given UTXOs, it depends on the fee rate. So maybe at a high fee rate both the change and `C` are negative, but at a lower fee rate they are not. Maybe today there's a fee spike or you are just in a hurry and use a high fee rate, tomorrow you can consolidate whatever is left at the min relay fee.

Reply:
Yes, that's clear. I was thinking in a situation where even at the `min_relay_fee` the utxos were uneconomical. But, would be probably fixed by the consolidation you mentioned in the last line.